src/libraries/System.IO.Compression/src/System/IO/Compression/ZipArchiveEntry.cs

```diff
@@ -964,7 +964,7 @@ private void WriteCrcAndSizesInLocalHeader(bool zip64HeaderUsed)
             long finalPosition = _archive.ArchiveStream.Position;
             BinaryWriter writer = new BinaryWriter(_archive.ArchiveStream);
 
-            bool zip64Needed = SizesTooLarge()
+            bool zip64Needed = SizesTooLarge() || (_offsetOfLocalHeader > uint.MaxValue)
```

Comment:
Should the additional condition be part of SizesTooLarge? It's not clear to me from a cursory look what happens if some parts of the code think zip64 is needed and others don't.